internal/mempool/mempool.go

```diff
@@ -1192,37 +1213,52 @@ func (mv *txMiningView) addTransaction(txDesc *mining.TxDesc, findTx txDescFind)
 	}
 }
 
-// addTransaction adds the passed transaction to the memory pool.  It should
-// not be called directly as it doesn't perform any validation.  This is a
-// helper for maybeAcceptTransaction.
-//
-// This function MUST be called with the mempool lock held (for writes).
-func (mp *TxPool) addTransaction(utxoView *blockchain.UtxoViewpoint, tx *dcrutil.Tx,
-	txType stake.TxType, height int64, fee int64, isTreasuryEnabled bool, totalSigOps int) {
+// newTxDesc returns a new TxDesc instance that captures mempool state
+// relevant to the provided transaction at the current time.
+func (mp *TxPool) newTxDesc(utxoView *blockchain.UtxoViewpoint, tx *dcrutil.Tx,
+	txType stake.TxType, prevHeight, bestHeight, fee int64, totalSigOps int) *TxDesc {
 
-	// Notify callback about vote if requested.
-	if mp.cfg.OnVoteReceived != nil && txType == stake.TxTypeSSGen {
-		mp.cfg.OnVoteReceived(tx)
+	var effectiveHeight int64
+	if prevHeight == blockHeightUnknown {
+		effectiveHeight = bestHeight
+	} else {
+		effectiveHeight = prevHeight
 	}
 
-	// Add the transaction to the pool and mark the referenced outpoints
-	// as spent by the pool.
 	msgTx := tx.MsgTx()
-	poolTxDesc := &TxDesc{
+	return &TxDesc{
 		TxDesc: mining.TxDesc{
 			Tx:          tx,
 			Type:        txType,
 			Added:       time.Now(),
-			Height:      height,
+			Height:      effectiveHeight,
```

Comment:
Yes, the reasoning behind persisting `Height` between the stage and main pool is for pruning tickets at a predictable height based on when it was added to the mempool (from a peer, for example). A reorg or disapproval that causes one of the ticket's parents to re-enter the mempool would push the ticket to the stage pool with same `Height` that it was initially seen with this change. Since the height is persisted when moving from the main pool to the stage pool, resetting the height when moving a staged ticket back to the main pool seemed like an odd behavior to me since the ticket never really left the mempool, depending on who's looking at it. I have no issue with walking this part back if it's cleaner to do so.

Reply:
Thanks, that makes sense and aligns with what I was assuming.

I don't see any issue with retaining the information from when the tx originally entered the mempool, but I have a few concerns with how it is being done:

- As mentioned in my prior comment, with this update, `Height` is retained from the time the tx originally entered the mempool, but other fields with the `TxDesc` are not (for example, `Added` will be the most recent time it was moved back to the main pool).  I feel that it should be consistent or otherwise should be well documented so that the context of the fields within the `TxDesc` is clear.
- Passing the `blockHeightUnknown` around feels like a bit of a hack.  If we do go the route or retaining some of the `TxDesc` info from when the tx originally entered the mempool, maybe it makes sense to just pass the `*TxDesc` for that tx instead.

It would probably be best to get @davecgh's thoughts before making an update on this though, since he is much more familiar with all of the edge cases involved.

Reply:
So the general notion of keeping the original height the transaction entered the mempool is sound, but I also agree with @rstaudt2 here that it really should apply to all relevant fields and be consistent.

In particular, I don't care for the `blockHeightUnknown` approach because it's yet another sentinel value (similar to `mining.UnminedHeight` which is also not something I care for) that I generally prefer to avoid.  The caller should have the information it needs to choose which one to send into the func properly versus sending in a sentinel value for the func itself to decide on.

That said, I don't see any particular issue here in terms of the actual observable effects over and above the issues that already exist in terms of transactions that leave and reenter the mempool in response to block changes.

While it is definitely not something to be addressed in this PR, something that I have really wanted to see for a while is the ability to apply a set of changes in one fell swoop (aka batches), as opposed to one at a time, and those changes would have special properties depending on the batch type.  There are several edge conditions, particular in reorganizations, which currently aren't really handled all that well, but it all works out due to the fact that they are quite infrequent and also have significant disincentives attached to them.

Reply:
This has been updated to simply move the original TxDesc reference between internal pools rather than creating a new object each time.